lib/src/renamer.dart

```diff
@@ -0,0 +1,259 @@
+// license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+import 'package:charcode/charcode.dart';
+import 'package:string_scanner/string_scanner.dart';
+
+/// Renamer is a small DSL for renaming things.
+///
+/// To create a Renamer, you provide a piece of code and a map of keys to
+/// functions that return some string for each key based on an input.
+///
+/// To rename, you pass some input to Renamer. It will return a string with the
+/// new name if the code provided a match, or null if no match was found for
+/// that input.
+///
+/// The language itself consists of a series of statements separated by
+/// semicolons or line breaks. Each statement has 3 clauses: the key clause,
+/// the matcher clause, and the output clause.
+///
+/// The key clause specifies which key to match on. This can be empty to refer
+/// to a key that's just the empty string.
+///
+/// The matcher clause is a regular expression that should match the entirety of
+/// the value of that key.
+///
+/// The output clause is the text that should be returned when the matcher
+/// succeeds and may contain references to captured groups in the matcher's
+/// regular expression.
+///
+/// For an input, statements are evaluated in order until a match is found. Once
+/// a match is found, its output is returned immediately, bypassing any
+/// subsequent statements.
+///
+/// These clauses are combined as `<key> <matcher> to <output>`. If the key in
+/// question is the empty string, you omit that clause, so you get
+/// `<matcher> to <output>`.
+///
+/// If you wish to include a semicolon or space in the matcher or output clause,
+/// you can either escape it with `\` or wrap the entire clause in single or
+/// double quotes.
+class Renamer<T> {
+  /// A map from keys to functions that take an input and return the value of
+  /// that key for that input.
+  final Map<String, String Function(T input)> keys;
+
+  /// The list of statements that are evaluated in order by this renamer.
+  final List<_Statement<T>> _statements;
+
+  Renamer._(this.keys, this._statements);
+
+  /// Creates a simple Renamer from [code] that only uses an empty key.
+  static Renamer<String> simple(String code) {
+    return Renamer(code, {'': (input) => input});
+  }
+
+  /// Creates a Renamer from [code] that takes a map from keys to string values
+  /// as input.
+  ///
+  /// [keys] is the list of keys that can be referenced in [code] and must
+  /// all appear in every input map.
+  static Renamer<Map<String, String>> map(String code, List<String> keys) {
+    return Renamer(code, {for (var key in keys) key: (input) => input[key]});
+  }
+
+  /// Creates a Renamer based on [code] and a map from keys to functions that
+  /// take an input and return a string.
+  ///
+  /// All keys must consist of only lowercase letters, underscores, and hyphens.
+  ///
+  /// If provided, [sourceUrl] will appear in parsing errors. It can be
+  /// a [String] or a [Uri].
```

Comment:
Added. This was necessary to provide some useful context to any exceptions that StringScanner throws (I pass `--rename` as the source "URL" so that any parsing exceptions there indicate that it's from the rename option, and not from an actual Sass file.